compiler/rustc_resolve/src/late.rs

```diff
@@ -3295,19 +3313,28 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             self.report_error(v.0, ResolutionError::VariableBoundWithDifferentMode(name, v.1));
         }
 
-        // 5) Finally bubble up all the binding maps.
-        maps
+        // 5) Bubble up the final binding map.
+        let is_never_pat = not_never_pats.is_empty();
```

Comment:
This is a bit... confusing IMO

So an or-pat is "never" if *all* of its constituents are never? This feels easy to mess up.

Reply:
Yeah, `Ok(x) | Err(!)` works anywhere a normal pattern would. Compare with `Ok(!) | Err(!)`, which is unreachable and doesn't take a body.

Unless you meant that the implementation is confusing?

Reply:
No, I just mean the invariants feel subtle because what makes the *product* of patterns a never pattern is if *one* is never, but what makes a *sum* of patterns a never pattern is if *all* are never.

I think this should be made more clear, preferably with documentation and examples, because we really shouldn't be making resolve more complicated without helping future readers out, lol.